test/uri/uri_parse_test.cc

```diff
@@ -0,0 +1,173 @@
+  });
+}
+
+TEST(URI_parse, success_2) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{
+        "http://[::1]:80/segment/index.html?query#frag"};
+    EXPECT_EQ(uri.recompose(), "http://::1:80/segment/index.html?query#frag");
+  });
+}
+
+TEST(URI_parse, success_3) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{
+        "http://user:pass@[::1]/segment/index.html?query#frag"};
+    EXPECT_EQ(uri.recompose(),
+              "http://user:pass@::1/segment/index.html?query#frag");
+  });
+}
+
+TEST(URI_parse, success_4) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{"http://user:pass@[::1]:80?query#frag"};
+    EXPECT_EQ(uri.recompose(), "http://user:pass@::1:80?query#frag");
+  });
+}
+
+TEST(URI_parse, success_5) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{
+        "http://user:pass@[::1]:80/segment/index.html#frag"};
+    EXPECT_EQ(uri.recompose(),
+              "http://user:pass@::1:80/segment/index.html#frag");
+  });
+}
+
+TEST(URI_parse, success_6) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{
+        "http://user:pass@[::1]:80/segment/index.html?query"};
+    EXPECT_EQ(uri.recompose(),
+              "http://user:pass@::1:80/segment/index.html?query");
+  });
+}
+
+TEST(URI_parse, success_7) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{"ftp://host:21/gnu/"};
+    EXPECT_EQ(uri.recompose(), "ftp://host:21/gnu/");
+  });
+}
+
+TEST(URI_parse, relative_1) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{"one/two/three"};
+    EXPECT_EQ(uri.recompose(), "one/two/three");
+  });
+}
+
+TEST(URI_parse, relative_2) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{"/one/two/three"};
+    EXPECT_EQ(uri.recompose(), "one/two/three");
+  });
+}
+
+TEST(URI_parse, relative_3) {
+  EXPECT_NO_THROW(
+      sourcemeta::jsontoolkit::URI uri{"//user:pass@localhost/one/two/three"});
+}
+
+TEST(URI_parse, real_life_1) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{
+        "http://sourceforge.net/projects/uriparser/"};
+    EXPECT_EQ(uri.recompose(), "http://sourceforge.net/projects/uriparser/");
+  });
+}
+
+TEST(URI_parse, real_life_2) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{
+        "http://sourceforge.net/project/platformdownload.php?group_id=182840"};
+    EXPECT_EQ(
+        uri.recompose(),
+        "http://sourceforge.net/project/platformdownload.php?group_id=182840");
+  });
+}
+
+TEST(URI_parse, mailto) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{"mailto:test@example.com"};
+    EXPECT_EQ(uri.recompose(), "mailto:test@example.com");
+  });
+}
+
+TEST(URI_parse, relative_4) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{"../../"};
+    EXPECT_EQ(uri.recompose(), "../../");
+  });
+}
+
+TEST(URI_parse, root_path) {
+  EXPECT_NO_THROW({
+    sourcemeta::jsontoolkit::URI uri{"/"};
+    EXPECT_EQ(uri.recompose(), "");
```

Comment:
Always doubtful on this one :/

Reply:
I don't remember very well, but on a previous PR, I think the conclusion was that `""` and `"/"` are different URIs, as `https://google.com` and `https://google.com/` are indeed different URIs

Reply:
I took a deeper look and I found interesting facts:

- `uriparser` test this case considering `"/"` as a [non empty segment](https://github.com/uriparser/uriparser/blob/bf0174e83164a4659c51c135399478bec389eafa/test/test.cpp#L121) that makes sens right? If there is a slash, there is an empty segment (at least). They use `pathTail` to check it. 
- When I parse `"/"` myself and that I explore our `internal` object I found this:
   ```text
   (lldb) p this->internal->uri
    (UriUriA) {
      scheme = (first = 0x0000000000000000, afterLast = 0x0000000000000000)
      userInfo = (first = 0x0000000000000000, afterLast = 0x0000000000000000)
      hostText = (first = 0x0000000000000000, afterLast = 0x0000000000000000)
      hostData = {
        ip4 = nullptr
        ip6 = nullptr
        ipFuture = (first = 0x0000000000000000, afterLast = 0x0000000000000000)
      }
      portText = (first = 0x0000000000000000, afterLast = 0x0000000000000000)
      pathHead = nullptr
      pathTail = nullptr
      query = (first = 0x0000000000000000, afterLast = 0x0000000000000000)
      fragment = (first = 0x0000000000000000, afterLast = 0x0000000000000000)
      absolutePath = 1
      owner = 1
      reserved = 0x0000000000000000
    }
   ``` 
   As you probably see, `pathTail` is `nullptr` hahaha!
   
 Two divergents facts! I think that if I figure out why I don't have this empty segment, I resolve the issue (more or less)

Reply:
Maybe a clue!

They use [`uriParseUriA`](https://github.com/uriparser/uriparser/blob/bf0174e83164a4659c51c135399478bec389eafa/test/test.cpp#L75) while we are using `uriParseSingleUriA`

Reply:
Ah, maybe. I'm not sure to be honest. At this point you likely know more about `uriparser` than I do. Keep up the good research and happy to pair debug if it helps